File: src/main/java/com/truveta/opentoken/attributes/BaseAttribute.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.attributes;

import java.util.ArrayList;
import java.util.List;
import com.truveta.opentoken.attributes.validation.AttributeValidator;
import com.truveta.opentoken.attributes.validation.NotNullOrEmptyValidator;

/**
 * A base implementation of the {@link Attribute} interface.
 * 
 * <p>
 * This class provides a default implementation of the
 * {@link Attribute#validate(String)}
 * method that validates the attribute value against a set of validation rules.
 * </p>
 * 
 * <p>
 * The default validation rules are:
 * <ul>
 * <li>Not null or empty</li>
 * </ul>
 * </p>
 */
public abstract class BaseAttribute implements Attribute {
    
    private final List<AttributeValidator> validationRules;

    protected BaseAttribute(List<AttributeValidator> validationRules) {
        ArrayList<AttributeValidator> ruleList = new ArrayList<>();
        ruleList.add(new NotNullOrEmptyValidator());
        ruleList.addAll(validationRules);
        this.validationRules = List.copyOf(ruleList);
    }

    /**
     * Validates the attribute value against a set of validation rules.
     */
    @Override
    public boolean validate(String value) {
        if (value == null) {
            return false;
        }

        return validationRules.stream().allMatch(rule -> rule.eval(value));
    }
}

File: lib/nodejs/opentoken/src/attributes/BaseAttribute.ts
/**
 * Copyright (c) Truveta. All rights reserved.
 */

import { Attribute } from './Attribute';
import { Validator } from './validation/Validator';

/**
 * Base class for all attributes.
 */
export abstract class BaseAttribute implements Attribute {
  protected name: string;
  protected aliases: string[];
  protected validators: Validator[];

  /**
   * Constructs a new BaseAttribute.
   *
   * @param name - The name of the attribute.
   * @param aliases - The aliases of the attribute.
   * @param validators - The validators for the attribute.
   */
  constructor(name: string, aliases: string[] = [], validators: Validator[] = []) {
    this.name = name;
    this.aliases = aliases;
    this.validators = validators;
  }

  getName(): string {
    return this.name;
  }

  getAliases(): string[]  {
    return this.aliases;
  }

  /**
   * Normalizes the attribute value.
   * Default implementation returns the value as-is.
   * Subclasses should override this method to provide custom normalization.
   *
   * @param value - The attribute value.
   * @returns The normalized attribute value.
   */
  normalize(value: string): string {
    return value;
  }

  /**
   * Validates the attribute value using all configured validators.
   *
   * @param value - The attribute value.
   * @returns true if the value passes all validators; false otherwise.
   */
  validate(value: string): boolean {
    if (!value || value.trim().length === 0) {
      return false;
    }

    // Run all validators
    for (const validator of this.validators) {
      if (!validator.validate(value)) {
        return false;
      }
    }

    return true;
  }
}
